gt/main.c

```diff
@@ -1409,6 +1409,43 @@ gt_process_unparsed_incoming_pkt(struct acl_search *acl4,
 	rte_pktmbuf_free(pkt);
 }
 
+static void
+return_message(struct gt_instance *instance)
+{
+	int ret;
+	size_t reply_len;
+	const char *reply_msg;
+	struct dynamic_config *dy_conf = get_dy_conf();
+	struct dy_cmd_entry *entry = mb_alloc_entry(&dy_conf->mb);
+
+	if (entry == NULL) {
+		GT_LOG(ERR, "gt: failed to send new lua update return to Dynamic config block at lcore %d\n",
+			dy_conf->lcore_id);
+		return;
+	}
+
+	reply_msg = luaL_checklstring(instance->lua_state, -1, &reply_len);
+
+	if (reply_len > RETURN_MSG_MAX_LEN) {
+		strncpy(entry->u.gt.return_msg, reply_msg, RETURN_MSG_MAX_LEN);
+		entry->u.gt.return_msg[RETURN_MSG_MAX_LEN - 1] = '\0';
+
+		GT_LOG(WARNING,
+			"gt: the return message length (%lu) exceeds the limit\n",
+			reply_len);
+	}
+
+	entry->op = GT_UPDATE_POLICY_RETURN;
+	strncpy(entry->u.gt.return_msg, reply_msg, reply_len);
```

Comment:
Bug: the `if (reply_len > RETURN_MSG_MAX_LEN) {` doesn't change `reply_len`, so this call of `strncpy()` is not safe. Moreover, `strncpy()` is call twice when the previous `if` body is executed.